cirq/sim/clifford_simulator.py

```diff
@@ -0,0 +1,749 @@
+        self.zs[:,q1] ^= self.zs[:,q2]
+            
+    def _rowsum(self,q1,q2):
+        ''' Implements the "rowsum" routine defined by Aaronson and Gottesman.
+        This multiplies the stabilizer in row q1 by the stabilizer in row q2. '''
+        def g(x1,z1,x2,z2):
+            if not x1 and not z1:
+                return 0
+            elif x1 and z1:
+                return int(z2)-int(x2)
+            elif x1 and not z1:
+                return int(z2)*(2*int(x2)-1)
+            else:
+                return int(x2)*(1-2*int(z2))
+            
+        r = 2*int(self.rs[q1])+2*int(self.rs[q2])
+        for j in range(self.n):
+            r += g(self.xs[q2,j],self.zs[q2,j],self.xs[q1,j],self.zs[q1,j])
+        
+        r %= 4
+        
+        self.rs[q1] = bool(r) 
+        
+        self.xs[q1,:] ^= self.xs[q2,:]
+        self.zs[q1,:] ^= self.zs[q2,:]
+    
+    def stabilizers(self):
+        ''' Returns the stabilizer generators of the state. These
+        are n operators {S_1,S_2,...,S_n} such thate S_i |psi> = |psi> '''
+        stabilizers = [] 
+        
+        for i in range(self.n,2*self.n):
+            stabilizer = "-" if self.rs[i] else ""
+            
+            for k in range(self.n):
+                if self.xs[i,k] & (not self.zs[i,k]):
+                    stabilizer += "X%d" % k
+                elif (not self.xs[i,k]) & self.zs[i,k]:
+                    stabilizer += "Z%d" % k
+                elif self.xs[i,k] & self.zs[i,k]:
+                    stabilizer += "Y%d" % k
+            
+            stabilizers.append(stabilizer)
+            
+        return stabilizers
+                
+    def _measure(self,q):
+        ''' Performs a projective measurement on the q'th qubit. 
+
+        Returns: the result (0 or 1) of the measurement.
+        '''
+        is_commuting = True
+        for i in range(self.n,2*self.n):
+            if self.xs[i,q]:
+                p = i
+                is_commuting = False
+                break
+                
+                
+        if is_commuting:
+            self.xs[2*self.n,:] = False
+            self.zs[2*self.n,:] = False
+            self.rs[2*self.n] = False
+            
+            for i in range(self.n):
+                if self.xs[i,q]:
+                    self._rowsum(2*self.n,self.n+i)
+            return int(self.rs[2*self.n])
+        
+        else:
+            for i in range(2*self.n):
+                if i != p and self.xs[i,q]:
+                    self._rowsum(i,p)
+
+            self.xs[p-self.n,:] = self.xs[p,:]
+            self.zs[p-self.n,:] = self.zs[p,:]
+            self.rs[p-self.n] = self.rs[p]
+
+            self.xs[p,:] = False
+            self.zs[p,:] = False
+
+            self.zs[p,q] = True
+
+            self.rs[p] = bool(np.random.randint(2))
+
+            return int(self.rs[p])
+
+
+class CH_Form():
+    '''
+    A representation of stabilizer states using the CH form,
+
+        |psi> = omega U_C U_H |s>
```

Comment:
You can use dollar signs to make this show up as latex in the docs.

Reply:
Done.